Source/WebCore/Modules/url-pattern/URLPattern.cpp

```
@@ -26,29 +26,205 @@
 #include "config.h"
 #include "URLPattern.h"
 
+#include "URLPatternCanonical.h"
 #include "URLPatternInit.h"
 #include "URLPatternOptions.h"
 #include "URLPatternResult.h"
 #include <wtf/RefCounted.h>
+#include <wtf/URL.h>
 
 namespace WebCore {
 
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(URLPattern);
 
+static String escapeStringHelper(StringView input)
+{
+    StringBuilder result;
+    char escapeCharacters[] = "\\+*?:{}()";
+
+    for (auto character : input.span8()) {
+        if (strchr(escapeCharacters, character))
+            result.append('\\');
+
+        result.append(character);
+    }
+
+    return result.toString();
+}
+
+static String processBaseURLString(String&& input, BaseURLStringType type)
+{
+    if (type != BaseURLStringType::Pattern)
+        return input;
+
+    return escapeStringHelper(input);
+}
+
+static URLPatternInit processInit(URLPatternInit&& init, BaseURLStringType type, String&& argProtocol, String&& argUsername,
+    String&& argPassword, String&& argHostname, String&& argPort, String&& argPathname, String&& argSearch, String&& argHash)
+{
+    URLPatternInit result;
+
+    if (!argProtocol.isEmpty())
+        result.protocol = argProtocol;
+
+    if (!argUsername.isEmpty())
+        result.username = argUsername;
+
+    if (!argPassword.isEmpty())
+        result.password = argPassword;
+
+    if (!argHostname.isEmpty())
+        result.hostname = argHostname;
+
+    if (!argPort.isEmpty())
+        result.port = argPort;
+
+    if (!argPathname.isEmpty())
+        result.pathname = argPathname;
+
+    if (!argSearch.isEmpty())
+        result.search = argSearch;
+
+    if (!argHash.isEmpty())
+        result.hash = argHash;
+
+    URL baseURL;
+
+    if  (!(init.baseURL.isEmpty())) {
+
+        baseURL = URL(init.baseURL);
+
+        if (!baseURL.isValid() || baseURL.isEmpty()) {
+            // FIXME // Throw Type Error Exception // Log error for now
+            WTFLogAlways("baseURL is not valid");
```

We cannot land a WTFLogAlways.

---

The spec says to throw an exception if baseURL fails parsing. Seems this function should return an `ExceptionOr<URLPatternInit>` and we should return an Exception here.